### ISSUE TITLE
Study of API Testing

### ISSUE THREAD

**Issue by ThunderSon** [OWASP/wstg#492]
**What would you like to happen?**
APIs come in different flavors, REST, SOAP, GraphQL, etc.
To simply say API Testing, that's a little bit ludicrous, and be broken down better. Issue #267 should adapt according to the plan that will be described in this issue.

What do you think? I believe this can handle a call, or maybe trying to plan it out and then combine our ideas in one thread to see what would work best.

Pinging the API Security team to get their feedback on this in order to have a well fleshed out plan
cc: @ErezYalon @PauloASilva @inonshk

**Comment by PauloASilva** [comment]
Thanks for pinging.
I do agree we can do a better job breaking the API Testing per "API flavor".

Since GraphQL is getting lots of attention, I suggest starting with this one, followed by REST and SOAP. I think we should leave the door open to other flavors such as JSON-RPC, ...

What's the plan?

Cheers,
Paulo A. Silva

**Comment by ThunderSon** [comment]
I agree with that, as GraphQL is the least documented attack vector between them.

What I have in mind is this:
1. Create the issues, with priorities (GQL -> REST -> SOAP)
2. Create an outline on the test scenario.
3. Divide the work between the possible attack vectors for every scenario in order to be tackled by contributors (whether it was from our side, or yours, if it's of interest to you).

In the issue opened, we'll need to provide references so that a contributor is capable of understanding on what their scenario will be based on.

What do you think?

**Comment by ThunderSon** [comment]
Work has started on the GraphQL Cheat Sheet.
Once that is done, I believe effort will come over to this side to write to the offensive part.

**Comment by ThunderSon** [comment]
An attack that was disclosed just today on Twitter's GraphQL:
https://hackerone.com/reports/885539

**Comment by PauloASilva** [comment]
Nice one.

We (me and @dsopas) will present and discuss two GraphQL issues we've found recently as part of our DefCon AppSec Village talk: [API (in)Security TOP 10: Guided tour to the Wild Wild World of APIs](https://www.appsecvillage.com/agenda/2020/api-in-security-top-10-guided-tour-to-the-wild-wild-world-of-apis).

Since DefCon is in Safe Mode, you'll be able to watch this and all other talks at [AppSec Village YouTube channel](https://t.co/lbo26XgfMj).

**Comment by PauloASilva** [comment]
Nice one.

We (me and @dsopas) will present and discuss two GraphQL issues we've found recently as part of our DefCon AppSec Village talk: [API (in)Security TOP 10: Guided tour to the Wild Wild World of APIs](https://www.appsecvillage.com/agenda/2020/api-in-security-top-10-guided-tour-to-the-wild-wild-world-of-apis).

Since DefCon is in Safe Mode, you'll be able to watch this and all other talks at [AppSec Village YouTube channel](https://www.youtube.com/c/appsecvillage).

**Comment by ThunderSon** [comment]
Sounds lovely! Looking forward to it :)

**Comment by PauloASilva** [comment]
On the API Testing side,  [MindAPI] is becoming a valuable asset. Not sure how it can be included, but I am sure it worth a reference.

Cheers, 
Paulo A. Silva

[MindAPI]: https://dsopas.github.io/MindAPI/play/

**Comment by mandalamr** [comment]
Hi, I’m new. Is this issue still active?

**Comment by kingthorin** [comment]
Yup still outstanding.

Make sure you read all the history and the same for any linked issues or PRs.

Probably best to provide a plan before starting work.

**Comment by PaarthPandey10** [comment]
Hi @kingthorin and team,

I’ve read through the history here and the related discussion in #351. Breaking this down by 'API flavor' makes perfect sense. Given my current background in networking and tech-legal compliance (which touches heavily on data structure and REST architectures), I’d like to propose a structured plan for tackling the REST API segment first, as it’s the most mature but often the most scattered in testing guides.

Proposed Action Plan for REST Testing Integration:

* Audit Existing Content: Review the current monolithic API testing sections and extract all REST-specific methodologies.
* Structure the REST Segment: Create an outline mirroring the WSTG template covering:
* Information Gathering (Endpoint discovery, parsing Swagger/OpenAPI docs)
* Authentication & Authorization (JWT manipulation, OAuth misconfigurations)
* Input Validation (Mass Assignment, HTTP Method tunneling)
* Draft & Review: Once the outline is approved, I can begin drafting the specific Markdown sections.

Does this approach for the REST vertical align with the team's vision for v5.0? If so, I’d be happy to own this segment and get an outline drafted for review.

**Comment by kingthorin** [comment]
@PaarthPandey10 we've kinda started API content as a separate section at this point, 4.12.

Unless you have a really compelling argument to spread it throughout existing content then I'd suggest basing your plan on enhancing/expanding 4.12.

**Comment by PaarthPandey10** [comment]
Thanks for the guidance, @kingthorin. That makes perfect sense. I will pivot the plan to focus exclusively on enhancing and expanding Section 4.12 (Testing_for_APIs.md).

I will begin extracting the relevant REST assessment methodologies from the legacy cheat sheet (REST_CS_Migrate.md) and weave them directly into the 4.12 section, ensuring strict adherence to the WSTG style guide. I'll drop a PR for review once the draft is ready.

**Comment by PaarthPandey10** [comment]
Hi @kingthorin,

I've opened PR #1427 which addresses this issue. As discussed, I migrated the core REST methodologies from the legacy cheat sheet and integrated them directly into the newly structured 12-API_Testing section (specifically the Reconnaissance, Data Exposure, and Function-Level Authorization files) to keep the content localized.

Looking forward to your review!

**Comment by kingthorin** [comment]
There's no need to update the issue with PR info. I get notifications on both. It'll probably take me a few days (if not longer) to get to. I didn't manage any reviews last week and there's two PRs ahead of you. Don't take any delay as lack of appreciation, it's just lack of resources :wink:

**Comment by kingthorin** [comment]
Grok’s assessment:

Testing_for_APIs.md content is fully superseded by 4.12.0 + the existing Authorization / Session chapters. No further migration required; the file can be removed.
